Auto save progress at the end of Chapter

shaavan opened the pull request:
This PR addresses part-1 of #480 

> Auto-save progress at the end of the chapter. Make it reusable so we can continue to add it as we add more chapters.


<h3>Description:</h3>

Once the user reaches the final outro of the page and has created an account earlier, this PR allows the progress to be updated to the next chapter's intro page.
This PR also updates the relevant code to be directed to the correct chapter's tag on the chapter selection screen.

benalleng commented on lib/progress.ts:
```diff
@@ -129,16 +129,46 @@ export const getLastUnlockedLessonPath = (userProgressKey: string): string => {
 
-  return result
+export const getNextChapterNumber = (userProgressKey: string): number => {
+  const curLessonNumber = parseInt(userProgressKey.slice(2, 3))
```
lets just rename this `currentLessonNumber` for clarity

shaavan replied:
Addressed in ad7d8e0775e0c4948e19f3e9f25d8b3844b82cf2

Thanks for the suggestion!

benalleng reviewed (changes requested):
small change but I tested and it works great!

shaavan commented:
Updated from 99b4ec3df9a6f7578c19c0eea5995bdfc7e0c004 to ad7d8e0775e0c4948e19f3e9f25d8b3844b82cf2 ([pr545.01](https://github.com/shaavan/saving-satoshi/commits/pr545.01) -> [pr545.02](https://github.com/shaavan/saving-satoshi/commits/pr545.02), [diff](https://github.com/shaavan/saving-satoshi/compare/pr545.01..pr545.02))
Addressed @benalleng suggestion

Update:

1. Renamed `curLessonNumber` -> `currentLessonNumber`

ecurrencyhodler commented:
Hey Shashwat! This works great. Quick question. Does this change also apply to auto-saving at the end of ch. 1 and 2? I tried going through ch. 1 with a brand new profile. I got to the outro page, didn't click "complete", and instead went straight to the chapters overview page and don't think it saved my progress.


https://github.com/saving-satoshi/saving-satoshi/assets/32662508/9f28cf51-fb24-4f1d-bb25-c5dfe0764b93

benalleng commented:
@ecurrencyhodler just for clarification `outro-1` isn't really the end of the chapter in that case. This logic will only work on `outro-2`

ecurrencyhodler commented:
Okay it works lol my bad.

shaavan commented:
> Okay it works lol my bad.

Haha! No problem! Thanks for the thorough testing, @ecurrencyhodler! :)

ecurrencyhodler commented:
I think @benalleng saw something he wanted to provide feedback on here. After you address that let's try to merge.

afterburn commented on lib/progress.ts:
```diff
@@ -129,16 +129,46 @@ export const getLastUnlockedLessonPath = (userProgressKey: string): string => {
+  const nextChapterIntro = 'CH' + getNextChapterNumber(userProgressKey) + 'INT1'
 
-  if (!result) {
+  if (keys[id + 1] && keys[id + 1] === nextChapterIntro) {
```
I'm not sure why on this line and on line 157 and I guess 161/162 we are adding either `1` or `2` to the `id`. Could you, perhaps with some comments, explain what it is you're doing here?

shaavan replied:
Addressed in 1b95577a705b2b9b43bde705c3454ffa009caf27
Thanks!

afterburn reviewed (changes requested):
Thanks @shaavan, left a comment.

benalleng commented:
> I think @benalleng saw something he wanted to provide feedback on here. After you address that let's try to merge.

It's ok, I just had some code I didn't understand but I think I'm good now
